nuxeo-core/nuxeo-core/src/main/java/org/nuxeo/ecm/core/api/AbstractSession.java

```diff
@@ -2087,18 +2087,17 @@ public void setRetainUntil(DocumentRef docRef, Calendar retainUntil, String comm
         checkPermission(doc, SET_RETENTION);
         Map<String, Serializable> options = new HashMap<>();
         options.put(CoreEventConstants.RETAIN_UNTIL, retainUntil);
-        String commentStart = retainUntil == null ? null : retainUntil.toInstant().toString();
-        if (comment == null) {
-            comment = commentStart;
-        } else if (commentStart != null) {
-            comment = commentStart + " " + comment;
-        }
-        options.put("comment", comment);
+        options.put("comment", retainUntil == null ? "" : retainUntil.toInstant().toString());
         DocumentModel docModel = readModel(doc);
-        notifyEvent(DocumentEventTypes.BEFORE_SET_RETENTION, docModel, options, null, null, true, false);
+        boolean currentIndeterminate = current == null || RETAIN_UNTIL_INDETERMINATE.compareTo(current) == 0;
+        String beforeRetentionEvent = currentIndeterminate ? DocumentEventTypes.BEFORE_SET_RETENTION
+                : DocumentEventTypes.BEFORE_EXTEND_RETENTION;
+        notifyEvent(beforeRetentionEvent, docModel, options, null, null, true, false);
         doc.setRetainUntil(retainUntil);
         docModel = readModel(doc);
-        notifyEvent(DocumentEventTypes.AFTER_SET_RETENTION, docModel, options, null, null, true, false);
+        String afterRetentionEvent = currentIndeterminate ? DocumentEventTypes.AFTER_SET_RETENTION
```

Comment:
Cool, this should solve the bug reported on https://jira.nuxeo.com/browse/NXP-28786

Reply:
done in the second comment :)